Add Code Review Process and User Experience, Housekeeping considerations

mattmenefee opened the pull request:
This commit is a work-in-progress.

mattmenefee committed:
Add Code Review Process and User Experience, Housekeeping considerations

This commit is a work-in-progress.

danivovich commented on README.md:
```diff
@@ -4,7 +4,35 @@ This is a living document. If you would like to contribute changes, please send
+
+- [ ] Have code reviewers been established for the project?
+
+  If it is a sole-developer project, is there an outside developer will bandwidth able to review proposed changes?
```
-will bandwidth- => with bandwidth

danivovich commented:
there is some great stuff here, thanks for contributing. nice insights

samwgoldman commented:
:100:  :heavy_multiplication_x: :+1:

mattmenefee committed:
Add more detail to the Feature Development Process guidelines

mattmenefee commented:
@samwgoldman I fleshed out some of the details of what I was thinking for the *Feature Development Process*. Does this error on the side of too much text? I'm pretty sure @yflicker would :-1:.

*Note:* that section is still a work-in-progress.

yflicker commented:
I think it looks great! I checked it out yesterday and today. Some thoughts / feedback below.

Can we set defaults for certain things? Like we've been trying to do with the prospectives? E.g. branch management and deployment process, process to signal a PR is ready to be merged, things like that for which we already have some established procedures in place.

If I understand correctly, Matt likes to do long-running (relatively speaking) WIP branches, others like to do smaller branches...different branching strategies. Do we create a README-matt.md or README-person.md to communicate one person's defaults vs. another's? Or are these just different tactics under "have a branch management strategy"? Or am I overthinking this?

Whatever the case, there's a lot (of good stuff) here, and when I think about how we'll process-ize this stuff and roll it out to the team, I think we should have some kind of prioritization or rollout strategy, so that we can work to perfect a few things at a time rather than everything all at once. But I'm open to whatever you all think is the best approach.

samwgoldman commented:
Hey, @mattmenefee. The section you added on development process is really good and valuable, but we're not quite ready to move on that yet. Once we wrap up the documentation on our PM process, we're going to tackle code management process, and that would be the perfect time to roll in the work you've done here.

Overall, I'm not convinced that this checklist is the right place to document code management process. Ideally, we'd document those choices in a prospective, the same way we document project management. It's more complex than a checkbox—it's selecting tactics and assigning responsibilities.

I think we should keep this checklist narrower in scope. Anyone who is actively reviewing a PR should be able to pull this up and run through it to gain confidence their feedback is complete. At that point, the dev process points you added should already be sorted out.

So, long story short, can you remove the development process stuff from this PR, and create a separate doc in drive so we can revisit once we're ready to document that aspect of our process?

The items you added to the code review section I want to merge in, because I think it's a great addition to this document.

Sound good?

mattmenefee commented:
@samwgoldman that is the plan. Fleshing out the details of that was helping me think of ideas for the lower section, so I'd like to continue working on each in parallel and try to get it all wrapped up this week. Or if you'd prefer, I can split out what I've contributed to the lower section and deliver that now, and then open up a new pull request if I come up with anything else for that section. Let me know.

mattmenefee committed:
Remove Feature Development Process information

Remove Feature Development Process information in favor of moving this to the Prospective template or the SmartLogic Standard Processes document.

mattmenefee committed:
Add step to check sample data during database migrations

This addresses #4.

mattmenefee commented:
@yflicker @danivovich @samwgoldman I moved the section on the development process to [a Google doc](https://docs.google.com/document/d/1zjNveVkrLrJ3edCRLOJTPQjDdDZzlJNPWXg9KBzNiHE/edit#) for now. Do you have any additional feedback? Please review and merge if appropriate.
